php/class-wp-js-widget.php

```diff
@@ -459,38 +494,167 @@ protected function render_input_attrs( $input_attrs ) {
+	 *     @type string $label       Field label, uses schema field `description` by default.
+	 *     @type string $class       Class name. Optional.
+	 *     @type string $placeholder Input placeholder, uses schema field `default` by default. Optional.
+	 *     @type string $min         Minimum, uses schema field `minimum` by default. Optional.
+	 *     @type string $max         Maximum, uses schema field `maximum` by default. Optional.
+	 *     @type string $help        Optional help text that appears below the input.
+	 * }
 	 */
 	protected function render_form_field_template( $args = array() ) {
-		$defaults = array(
-			'name' => '',
-			'label' => '',
+		$item_schema = $this->get_item_schema();
+		$default_input_attrs = array(
 			'type' => 'text',
-			'choices' => array(),
-			'value' => '',
-			'placeholder' => '',
-			'help' => '',
+			'class' => '',
 		);
+
+		if ( ! empty( $args['name'] ) ) {
+			_deprecated_argument( __FUNCTION__, '0.3.0', __( 'The args[name] param is deprecated in favor of args[field].', 'js-widgets' ) );
+
+			if ( empty( $args['field'] ) && ! empty( $item_schema[ $args['name'] ] ) ) {
+				$args['field'] = $args['name'];
+			}
+		}
+
+		$field_name = ! empty( $args['field'] ) ? $args['field'] : null;
+		if ( $field_name && ! empty( $item_schema[ $field_name ] ) ) {
+			$field_schema = $item_schema[ $field_name ];
+			$schema_to_input_attrs_mapping = array(
+				'description' => 'label',
+				'minimum' => 'min',
+				'maximum' => 'max',
+				'default' => 'placeholder',
+			);
+			foreach ( $schema_to_input_attrs_mapping as $schema_key => $input_attr_key ) {
+				if ( isset( $field_schema[ $schema_key ] ) ) {
+					$default_input_attrs[ $input_attr_key ] = $field_schema[ $schema_key ];
+				}
+			}
+			if ( isset( $field_schema['properties']['raw'] ) ) {
+				foreach ( $schema_to_input_attrs_mapping as $schema_key => $input_attr_key ) {
+					if ( isset( $field_schema['properties']['raw'][ $schema_key ] ) ) {
+						$default_input_attrs[ $input_attr_key ] = $field_schema['properties']['raw'][ $schema_key ];
+					}
+				}
+			}
+
+			if ( isset( $field_schema['type'] ) && is_string( $field_schema['type'] ) ) {
+				$schema_type = $field_schema['type'];
+
+				if ( 'boolean' === $schema_type ) {
+					$default_input_attrs['type'] = 'checkbox';
+				} elseif ( 'integer' === $schema_type || 'number' === $schema_type ) {
+					$default_input_attrs['type'] = 'number';
+				} elseif ( 'string' === $schema_type && isset( $field_schema['format'] ) ) {
+					if ( 'uri' === $field_schema['format'] ) {
+						$default_input_attrs['type'] = 'url';
+					} elseif ( 'email' === $field_schema['format'] ) {
+						$default_input_attrs['type'] = 'email';
+					}
+					// @todo Support date-time format.
+				}
+
+				if ( 'integer' === $schema_type ) {
+					$default_input_attrs['step'] = '1';
+				}
+			}
+
+			if ( isset( $field_schema['enum'] ) ) {
```

Comment:
Perhaps. It could map to a `select` or a `radio`. But since the `choices` would have to be supplied in any case (to get the labels), to me it seemed fine to just leave it so that the user has to explicitly pass the field `type` in this case.